File: client/src/api/matchaLitePricing.ts
import { useEffect, useState } from 'react'
import { api } from './client'

// Product codes priced through this table: the two Lite variants that share
// the /lite/signup page + checkout (standard, with an employee roster;
// Essentials, no roster/no OSHA logs), standalone Matcha Compliance, and the
// Lite add-ons (block_size=1 ⇒ per-employee/month) — each priced independently.
export type MatchaLiteProductCode =
  | 'matcha_lite'
  | 'matcha_lite_essentials'
  | 'matcha_compliance'
  | 'addon_voice_intake'
  | 'addon_hris_sync'
  | 'addon_handbook_watch'

export interface MatchaLitePricingConfig {
  price_per_block_cents: number
  block_size: number
  effective_price_per_block_cents: number
  sale_active: boolean
  min_headcount: number
  max_headcount: number
}

export function fetchMatchaLitePricing(productCode: MatchaLiteProductCode = 'matcha_lite') {
  return api.get<MatchaLitePricingConfig>(`/resources/matcha-lite/pricing?product_code=${encodeURIComponent(productCode)}`)
}

export function computeLitePriceDollars(headcount: number, pricing: MatchaLitePricingConfig): number {
  return Math.ceil(headcount / pricing.block_size) * (pricing.effective_price_per_block_cents / 100)
}

// Live Matcha Lite price, fetched once per productCode. Callers should fall
// back to a sane default (e.g. 300) for max_headcount while `pricing` is null.
export function useMatchaLitePricing(productCode: MatchaLiteProductCode = 'matcha_lite'): MatchaLitePricingConfig | null {
  const [pricing, setPricing] = useState<MatchaLitePricingConfig | null>(null)
  useEffect(() => {
    setPricing(null)
    fetchMatchaLitePricing(productCode).then(setPricing).catch(() => {})
  }, [productCode])
  return pricing
}

// Admin — Matcha Lite pricing config (base + sale price). Distinct shape from
// MatchaLitePricingConfig above: no computed effective_price_per_block_cents,
// but exposes the raw sale price + change metadata for editing.
export interface MatchaLitePricingAdminConfig {
  price_per_block_cents: number
  block_size: number
  sale_price_per_block_cents: number | null
  sale_active: boolean
  min_headcount: number
  max_headcount: number
  updated_at: string | null
  updated_by: string | null
}

export interface MatchaLitePricingUpdate {
  price_per_block_cents: number
  block_size: number
  sale_price_per_block_cents: number | null
  sale_active: boolean
  min_headcount: number
  max_headcount: number
}

export function fetchMatchaLitePricingAdmin(productCode: MatchaLiteProductCode = 'matcha_lite') {
  return api.get<MatchaLitePricingAdminConfig>(`/admin/matcha-lite-pricing?product_code=${encodeURIComponent(productCode)}`)
}

export function saveMatchaLitePricingAdmin(
  update: MatchaLitePricingUpdate,
  productCode: MatchaLiteProductCode = 'matcha_lite',
) {
  return api.put<MatchaLitePricingAdminConfig>(`/admin/matcha-lite-pricing?product_code=${encodeURIComponent(productCode)}`, update)
}

File: server/app/core/services/matcha_lite_pricing.py
"""Matcha Lite pricing — DB-backed, admin-configurable via /admin/matcha-lite-pricing.

Price is a step function: ceil(headcount / block_size) * price_per_block, with
an optional sale override. Supersedes stripe_service.matcha_lite_price_cents
for the actual checkout path — that function stays in place for Matcha-X,
which is unchanged, still a hardcoded stub.

Matcha Compliance also uses this table (product_code='matcha_compliance'),
seeded flat at $8/head (block_size=1) — see migration mlpricing03.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

from ...database import get_connection

SELECT_COLUMNS = """
    price_per_block_cents, block_size, sale_price_per_block_cents,
    sale_active, min_headcount, max_headcount, updated_at, updated_by
"""

# signup_source values priced via this table: the two Lite variants (share the
# /lite/signup page + /checkout/lite endpoint) plus standalone Matcha Compliance
# (/compliance/signup + /checkout/compliance) — each its own row.
PRODUCT_CODES = ("matcha_lite", "matcha_lite_essentials", "matcha_compliance")

_FALLBACK_DEFAULTS = {
    "matcha_lite": dict(price_per_block_cents=5000, block_size=10, min_headcount=1, max_headcount=300),
    "matcha_lite_essentials": dict(price_per_block_cents=4000, block_size=10, min_headcount=1, max_headcount=300),
    "matcha_compliance": dict(price_per_block_cents=800, block_size=1, min_headcount=1, max_headcount=300),
}


@dataclass
class MatchaLitePricing:
    price_per_block_cents: int
    block_size: int
    sale_price_per_block_cents: Optional[int]
    sale_active: bool
    min_headcount: int
    max_headcount: int
    updated_at: Optional[str]
    updated_by: Optional[str]

    @property
    def effective_price_per_block_cents(self) -> int:
        if self.sale_active and self.sale_price_per_block_cents is not None:
            return self.sale_price_per_block_cents
        return self.price_per_block_cents


def row_to_pricing(row) -> MatchaLitePricing:
    return MatchaLitePricing(
        price_per_block_cents=row["price_per_block_cents"],
        block_size=row["block_size"],
        sale_price_per_block_cents=row["sale_price_per_block_cents"],
        sale_active=row["sale_active"],
        min_headcount=row["min_headcount"],
        max_headcount=row["max_headcount"],
        updated_at=row["updated_at"].isoformat() if row["updated_at"] else None,
        updated_by=row["updated_by"],
    )


async def get_matcha_lite_pricing(conn=None, product_code: str = "matcha_lite") -> MatchaLitePricing:
    """Fetch the live pricing config for `product_code` ('matcha_lite' or
    'matcha_lite_essentials' — see PRODUCT_CODES).

    Pass an existing `conn` to reuse a connection already held by the caller
    (e.g. the checkout endpoint, which also reads headcount in the same
    request); otherwise a short-lived connection is opened.
    """
    query = f"SELECT {SELECT_COLUMNS} FROM matcha_lite_pricing WHERE product_code = $1"

    if conn is not None:
        row = await conn.fetchrow(query, product_code)
    else:
        async with get_connection() as c:
            row = await c.fetchrow(query, product_code)

    if row is None:
        # Defensive fallback if the seed row was somehow deleted — matches
        # the launch defaults (migrations mlpricing01/mlpricing02).
        defaults = _FALLBACK_DEFAULTS.get(product_code, _FALLBACK_DEFAULTS["matcha_lite"])
        return MatchaLitePricing(
            sale_price_per_block_cents=None,
            sale_active=False,
            updated_at=None,
            updated_by=None,
            **defaults,
        )
    return row_to_pricing(row)


def compute_matcha_lite_price_cents(pricing: MatchaLitePricing, headcount: int) -> Optional[int]:
    """Monthly price in cents for `headcount` under `pricing`, or None if out of range."""
    if headcount < pricing.min_headcount or headcount > pricing.max_headcount:
        return None
    blocks = math.ceil(headcount / pricing.block_size)
    return blocks * pricing.effective_price_per_block_cents
